frank/fit.py

```diff
@@ -354,21 +358,27 @@ def perform_fit(u, v, vis, weights, geom, model):
         model['plotting']['diag_plot']
 
     t1 = time.time()
-    FF = radial_fitters.FrankFitter(Rmax=model['hyperpriors']['rout'],
-                                    N=model['hyperpriors']['n'],
+    FF = radial_fitters.FrankFitter(Rmax=model['hyperparameters']['rout'],
+                                    N=model['hyperparameters']['n'],
                                     geometry=geom,
-                                    alpha=model['hyperpriors']['alpha'],
-                                    weights_smooth=model['hyperpriors']['wsmooth'],
-                                    tol=model['hyperpriors']['iter_tol'],
-                                    max_iter=model['hyperpriors']['max_iter'],
+                                    alpha=model['hyperparameters']['alpha'],
+                                    weights_smooth=model['hyperparameters']['wsmooth'],
+                                    tol=model['hyperparameters']['iter_tol'],
+                                    max_iter=model['hyperparameters']['max_iter'],
                                     store_iteration_diagnostics=need_iterations
                                     )
 
     sol = FF.fit(u, v, vis, weights)
 
+    if model['hyperparameters']['nonnegative']:
+        # Replace the mean solution with the best fit nonnegative one
+        logging.info('  `nonnegative` is `true` in your parameter file --> Replacing `sol.mean` with the best fit nonnegative profile')
+        setattr(sol, '_mu', sol.solve_non_negative())
+        model['input_output']['save_prefix'] += '_nonnegative'
+
```

Comment:
86093bf adds the attribute `_nonneg` to the `sol` object if `model['hyperparameters']['nonnegative']` is True, separately retaining `_mu` as an attribute of `sol`.

Reply:
I'm still not content with this solution. There are a few bad design choices here:

1. Methods / variables beginning with an `_` are meant to be private - they should not be used by functions / classes outside of the class itself.
2. You are changing the nature of the object. Someone loading in a sol object saved with fit.py might find a different structure to any they can create themselves, which could be confusing.

Since the sol object has the solve_non_negative method, why not just call that where it is needed? The function is not particularly expensive. If you're genuinely worried about the cost of solving for the non-negative solution multiple times, then the solution is to add caching to the solve_non_negative function.